woocommerce/woocommerce-ios
# Update Settings > Beta Features for Products teaser release

Fixes #1508 

## Changes
- In `BetaFeaturesViewController`, added Products visibility section with a switch cell and a description cell
  - Only showed the stats v4 section if the site is eligible
- In `SettingsViewController`, conditionally showed the Experimental Features row with feature flags and whether the site is eligible for stats v4
- In Yosemite layer, added `loadProductsVisibility` and `setProductsVisibility` actions to `AppSettingsStore`. Created `ProductsVisibilityPListWrapper` for internal storage as a plist
  - Testing: some basic unit tests, and moved `MockInMemoryStorage` to an internal mock for reuse
- In `MainTabBarController`, listened to the Products visibility change notification from `BetaFeaturesViewController`
  - Originally I wanted to pass a callback from `MainTabBarController` to `DashboardViewController` to `SettingsViewController` to `BetaFeaturesViewController`, however the `SettingsViewController` is initialized from the Storyboard so I wasn't able to access `SettingsViewController` in `DashboardViewController`. Eventually, I used the less ideal `NotificationCenter` to notify the Products visibility change from the Beta Features to the Main Tab Bar
  - Because we show an Orders badge and a purple dot on the Orders and Reviews tab respectively, I had to refactor how the badge/dot is shown/hidden with the tab index which is now variable
  - I also noticed that `UITabBar`'s subviews could be in any order, so I created a `UITabBar` extension to return the ordered actionable subviews for locating the tab bar button by tab index

## Testing

- Launch the app --> Products tab should not be shown 
- Go to Settings --> should see a row "Experimental Features"
- Tap on "Experimental Features"
- Turn on the Products switch --> the Products tab should appear
- Relaunch the app --> Products tab should still be shown
- Go to Settings --> should see a row "Experimental Features"
- Tap on "Experimental Features"
- Turn off the Products switch --> the Products tab should disappear
- Relaunch the app --> Products tab should still be hidden

## Example screenshots

Products on | Products off
-- | --
![Simulator Screen Shot - iPhone 8 - 2019-11-22 at 12 01 44](https://user-images.githubusercontent.com/1945542/69402573-33b1b980-0d33-11ea-9229-b2641c75f850.png) | ![Simulator Screen Shot - iPhone 8 - 2019-11-22 at 12 01 46](https://user-images.githubusercontent.com/1945542/69402575-344a5000-0d33-11ea-9190-89e1eb189142.png)


Update release notes:
